internal/v4/archive.go

```
@@ -208,3 +213,52 @@ func (h *handler) openBlob(id *charm.Reference) (blobstore.ReadSeekCloser, int64
 	}
 	return r, size, nil
 }
+
+func (h *handler) bundleCharms(ids []string) (map[string]charm.Charm, error) {
+	charms := make(map[string]charm.Charm, len(ids))
+	for _, id := range ids {
+		url, err := charm.ParseReference(id)
+		if err != nil {
+			// Ignore this error. This will be caught in the bundle
+			// verification process and will be returned to the user
+			// along with other bundle errors.
+			continue
+		}
+		if err = h.resolveURL(url); err != nil {
+			if errgo.Cause(err) == params.ErrNotFound {
+				// Ignore this error too, for the same reasons
+				// described above.
+				continue
+			}
+			return nil, err
+		}
+		r, size, err := h.openBlob(url)
+		if err != nil {
+			if err == params.ErrNotFound {
+				// Same as above: do not include this charm in charms.
+				// The bundle verification process will complain later.
+				continue
+			}
+			return nil, err
+		}
+		defer r.Close()
+		ch, err := charm.ReadCharmArchiveFromReader(&readerAtSeeker{r}, size)
+		if err != nil {
+			return nil, err
+		}
+		charms[id] = ch
+	}
+	return charms, nil
+}
+
+// verificationError returns an error whose string representation includes all
+// the verification error messages stored in err.
+// Note that err must be a *charm.VerificationError.
+func verificationError(err error) error {
```

I'm not _entirely_ sure we want to return all the error messages here. Or rather, if we
do return all the error messages, it would be good if they were in a form that was
easy for a client to split into separate messages for display. I'm not sure that
we guarantee that the error messages in VerificationError.Errors do not contain
semicolons.

My first thought was that we could just return the single error, and leave the
multiple errors to be printed by a client-side tool, run independently of the
charm store.

My second thought is that we could pack all the errors into JSON and
have that as part of the returned error message.

i.e. the returned error message would look like:

```
bundle verification error: ["charm \"wrdpress\" not found","another error etc"]
```

Then it's easy for a client to split them out again.

What do you think?